File: Python/2-Medium/00054._螺旋矩阵/00054._螺旋矩阵.py
def spiralOrder(matrix):
    """
    按顺时针螺旋顺序返回矩阵中的所有元素。

    Args:
    matrix (List[List[int]]): 一个二维整数数组。

    Returns:
    List[int]: 螺旋顺序的元素列表。
    """
    # 初始化结果列表和边界
    result = []
    if not matrix or not matrix[0]:
        return result
    top, bottom = 0, len(matrix) - 1
    left, right = 0, len(matrix[0]) - 1
    
    while left <= right and top <= bottom:
        # 遍历上边界
        for i in range(left, right + 1):
            result.append(matrix[top][i])
        top += 1
        
        # 遍历右边界
        for i in range(top, bottom + 1):
            result.append(matrix[i][right])
        right -= 1
        
        if top <= bottom:
            # 遍历下边界
            for i in range(right, left - 1, -1):
                result.append(matrix[bottom][i])
            bottom -= 1
        
        if left <= right:
            # 遍历左边界
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
            left += 1
    
    return result

# 测试代码
matrix = [[1,2,3],[4,5,6],[7,8,9]]
print(spiralOrder(matrix)) # 应输出: [1,2,3,6,9,8,7,4,5]

# 测试边界情况
matrix_empty = []
print(spiralOrder(matrix_empty)) # 应输出: []

matrix_one_row = [[1, 2, 3, 4]]
print(spiralOrder(matrix_one_row)) # 应输出: [1, 2, 3, 4]

matrix_one_col = [[1], [2], [3], [4]]
print(spiralOrder(matrix_one_col)) # 应输出: [1, 2, 3, 4]

File: C++/2-Medium/00054._螺旋矩阵/00054._螺旋矩阵.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> spiralOrder(vector<vector<int>>& matrix) {
    vector<int> result;
    if (matrix.empty()) return result; // 处理空矩阵的情况

    int m = matrix.size(), n = matrix[0].size(); // m和n分别为矩阵的行数和列数
    int left = 0, right = n - 1, top = 0, bottom = m - 1; // 初始化左、右、上、下四个边界

    while (left <= right && top <= bottom) {
        // 从左到右遍历上边界
        for (int column = left; column <= right; column++) {
            result.push_back(matrix[top][column]);
        }
        // 上边界向下缩进
        top++;

        // 从上到下遍历右边界
        for (int row = top; row <= bottom; row++) {
            result.push_back(matrix[row][right]);
        }
        // 右边界向左缩进
        right--;

        if (left <= right && top <= bottom) {
            // 从右到左遍历下边界
            for (int column = right; column >= left; column--) {
                result.push_back(matrix[bottom][column]);
            }
            // 下边界向上缩进
            bottom--;

            // 从下到上遍历左边界
            for (int row = bottom; row >= top; row--) {
                result.push_back(matrix[row][left]);
            }
            // 左边界向右缩进
            left++;
        }
    }
    return result;
}

int main() {
    vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};
    vector<int> result = spiralOrder(matrix);
    for (int i : result) {
        cout << i << " ";
    }
    cout << endl;
    return 0;
}